src/firmware/app/control/trajectory_planner.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include "firmware/shared/math/polynomial_2d.h"
+#include "firmware/shared/math/vector_2d.h"
+
+#define TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS                                              \
+    6000  // The maximum size of the array containing trajectory elements
+#define __TRAJECTORY_PLANNER_MAX_PATH_LENGTH__ 18  // [meters]
+
+// Struct that defines a single point on a trajectory
+// Includes the Position,and Timedata corresponding to that point
+typedef struct TrajectoryElement
+{
+    Vector2d_t position;
+    double time;
+} TrajectoryElement_t;
+
+typedef struct Trajectory
+{
+    TrajectoryElement_t* trajectory_elements;
+    unsigned int num_elements;
+} Trajectory_t;
+
+typedef struct FirmwareRobotPathParameters
+{
+    Polynomial2dOrder3_t path;
+    double t_start;
+    double t_end;
+    unsigned int num_segments;
```

Comment:
Just trying to understand the interaction between `FirmwareRobotPathParameters.num_segments`, `Trajectory.num_elements`, and `TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS`. Got a few suggestions here, but  let me know if they make sense or not:

- Remove `FirmwareRobotPathParameters.num_segments`, just use `Trajectory.num_elements`. This way we only have one source for the size, so conflicts are not possible. I know `Trajectory` is technically an _output_, but since we're allocating memory and passing it in it seems really easy for someone to just give an incorrect `num_segments` that exceeds the passed in memory. I know we have this warning: `THE NUMBER OF SEGMENTS MUST BE UNDER TRAJECTORY_PLANNER_MAX_NUMER_SEGMENTS`, but this only works if `trajectory_elements` is smaller then `TRAJECTORY_PLANNER_MAX_NUMER_SEGMENTS`, which it's not guaranteed to be.
- inside functions, allocate arrays using `Trajectory.num_elements` as the size instead of `TRAJECTORY_PLANNER_MAX_NUMER_SEGMENTS`. The reason that it's possible to allocate a dynamic length array like this is called [VLA](https://en.wikipedia.org/wiki/Variable-length_array)

With the above two suggestions, hopefully there should only ever be a single source of information for the number of segments on the trajectory, and it should be much easier to actually make sure we're using these functions in a safe manner.

Reply:
See in-person discussion

Reply:
Could we move the per-member comments to right above each member? ie.
```
// The path parameter value indicating the beginning of the considered path
float t_start;
// The path parameter value indicating the end of the considered path
float t_end;
...
```

Reply:
Move comments into the struct definition